internal/resource/registry.go

```diff
@@ -0,0 +1,58 @@
+package resource
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+type Registry interface {
+	// Register the given resource type and its hooks.
+	Register(reg Registration)
+
+	// Resolve the given resource type and its hooks.
+	Resolve(typ *pbresource.Type) (reg Registration, ok bool)
+}
+
+type Registration struct {
+	// Type is the GVK of the resource type.
+	Type *pbresource.Type
+
+	// In the future, we'll add hooks, the controller etc. here.
+}
+
+// Resource type registry
+type TypeRegistry struct {
+	registrations map[*pbresource.Type]Registration
```

Comment:
Could we guard this with a `sync.RWMutex`, please? Theoretically, everything should be registered on-startup (via package init functions) but we've been burned by gRPC's thread-unsafe resolver and balancer registries.

Reply:
👍 Fixed in https://github.com/hashicorp/consul/pull/16622/commits/66818c72de789b3305fbcec596602fdcb56d0a52